src/main/java/org/jfrog/hudson/pipeline/common/executors/CollectEnvExecutor.java

```diff
@@ -0,0 +1,46 @@
+package org.jfrog.hudson.pipeline.common.executors;
+
+import hudson.EnvVars;
+import hudson.FilePath;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.remoting.VirtualChannel;
+import jenkins.MasterToSlaveFileCallable;
+import org.jfrog.hudson.pipeline.common.types.buildInfo.Env;
+
+import java.io.File;
+import java.io.IOException;
+
+public class CollectEnvExecutor implements Executor {
+    private transient TaskListener listener;
+    private transient FilePath ws;
+    private Run build;
+    private Env env;
+    private EnvVars envVars;
+
+    public CollectEnvExecutor(Run build, TaskListener listener, FilePath ws, Env env, EnvVars envVars) {
+        this.listener = listener;
+        this.ws = ws;
+        this.env = env;
+        this.envVars = envVars;
+        this.build = build;
+    }
+
+    public void execute() throws IOException, InterruptedException {
+        env.append(ws.act(new CollectEnvExecutor.CollectEnvCallable(env.collectVariables(build, listener), envVars)));
```

Comment:
Due to the complexity of this line, I'm suggesting to split it to 2 lines:
The first line contains the collect env logic that runs in the agent.
The second line contains the collect build parameters logic that runs in the master.

Please add comments explaining rigorously each line.